From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nagarjuna Kristam <nkristam@nvidia.com>
Date: Wed, 16 Nov 2022 11:57:02 +0530
Subject: [PATCH 238/254] NVIDIA: SAUCE: PCI: tegra194: Add HSI support

Enable PCIe HSI faults and LIC interrupts for each error. Clear error
status and report EPL error to FSI.

Bug 3868928

Change-Id: I3984a420b3de421d1f08731cb6e3ff0831d3084f
Signed-off-by: Nagarjuna Kristam <nkristam@nvidia.com>
Signed-off-by: Manikanta Maddireddy <mmaddireddy@nvidia.com>
Reviewed-by: Bitan Biswas <bbiswas@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2898527
Tested-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>

diff --git a/drivers/pci/controller/dwc/pcie-tegra194.c b/drivers/pci/controller/dwc/pcie-tegra194.c
index 7f6641d747a25..ec5404a918a42 100644
--- a/drivers/pci/controller/dwc/pcie-tegra194.c
+++ b/drivers/pci/controller/dwc/pcie-tegra194.c
@@ -29,6 +29,12 @@
 #include <linux/random.h>
 #include <linux/reset.h>
 #include <linux/resource.h>
+#if IS_ENABLED(CONFIG_TEGRA_EPL)
+#include <linux/tegra-epl.h>
+#endif
+#if (IS_ENABLED(CONFIG_TEGRA_HSIERRRPTINJ))
+#include <linux/tegra-hsierrrptinj.h>
+#endif
 #include <linux/types.h>
 #include "pcie-designware.h"
 #include <soc/tegra/bpmp.h>
@@ -59,17 +65,34 @@
 #define APPL_INTR_EN_L0_0_LINK_STATE_INT_EN	BIT(0)
 #define APPL_INTR_EN_L0_0_MSI_RCV_INT_EN	BIT(4)
 #define APPL_INTR_EN_L0_0_INT_INT_EN		BIT(8)
+#define APPL_INTR_EN_L0_0_TLP_ERR_INT_EN	BIT(11)
+#define APPL_INTR_EN_L0_0_RASDP_INT_EN		BIT(12)
+#define APPL_INTR_EN_L0_0_PARITY_ERR_INT_EN	BIT(14)
 #define APPL_INTR_EN_L0_0_PCI_CMD_EN_INT_EN	BIT(15)
+#define APPL_INTR_EN_L0_0_PEX_RST_INT_EN	BIT(16)
 #define APPL_INTR_EN_L0_0_CDM_REG_CHK_INT_EN	BIT(19)
+#define APPL_INTR_EN_L0_0_SAFETY_UNCORR_INT_EN	BIT(20)
 #define APPL_INTR_EN_L0_0_SYS_INTR_EN		BIT(30)
 #define APPL_INTR_EN_L0_0_SYS_MSI_INTR_EN	BIT(31)
 
 #define APPL_INTR_STATUS_L0			0xC
 #define APPL_INTR_STATUS_L0_LINK_STATE_INT	BIT(0)
 #define APPL_INTR_STATUS_L0_INT_INT		BIT(8)
+#define APPL_INTR_STATUS_L0_TLP_ERR_INT		BIT(11)
+#define APPL_INTR_STATUS_L0_RASDP_INT		BIT(12)
+#define APPL_INTR_STATUS_L0_PARITY_ERR_INT	BIT(14)
 #define APPL_INTR_STATUS_L0_PCI_CMD_EN_INT	BIT(15)
 #define APPL_INTR_STATUS_L0_PEX_RST_INT		BIT(16)
 #define APPL_INTR_STATUS_L0_CDM_REG_CHK_INT	BIT(18)
+#define APPL_INTR_STATUS_L0_SAFETY_CORR_INT	BIT(19)
+#define APPL_INTR_STATUS_L0_SAFETY_UNCORR_INT	BIT(20)
+
+#define APPL_FAULT_EN_L0			0x10
+#define APPL_FAULT_EN_L0_TLP_ERR_FAULT_EN	BIT(11)
+#define APPL_FAULT_EN_L0_RASDP_FAULT_EN		BIT(12)
+#define APPL_FAULT_EN_L0_PARITY_ERR_FAULT_EN	BIT(14)
+#define APPL_FAULT_EN_L0_CDM_REG_CHK_FAULT_EN	BIT(18)
+#define APPL_FAULT_EN_L0_SAFETY_UNCORR_FAULT_EN	BIT(20)
 
 #define APPL_INTR_EN_L1_0_0				0x1C
 #define APPL_INTR_EN_L1_0_0_LINK_REQ_RST_NOT_INT_EN	BIT(1)
@@ -102,12 +125,44 @@
 
 #define APPL_INTR_STATUS_L1_9			0x54
 #define APPL_INTR_STATUS_L1_10			0x58
+
+#define APPL_FAULT_EN_L1_11			0x5c
+#define APPL_FAULT_EN_L1_11_NF_ERR_FAULT_EN	BIT(2)
+#define APPL_FAULT_EN_L1_11_F_ERR_FAULT_EN	BIT(1)
+
+#define APPL_INTR_EN_L1_11			0x60
+#define APPL_INTR_EN_L1_11_NF_ERR_INT_EN	BIT(2)
+#define APPL_INTR_EN_L1_11_F_ERR_INT_EN		BIT(1)
+
 #define APPL_INTR_STATUS_L1_11			0x64
+#define APPL_INTR_STATUS_L1_11_NF_ERR_STATE	BIT(2)
+#define APPL_INTR_STATUS_L1_11_F_ERR_STATE	BIT(1)
+
+#define APPL_FAULT_EN_L1_12			0x68
+#define APPL_FAULT_EN_L1_12_SLV_RASDP_ERR	BIT(1)
+#define APPL_FAULT_EN_L1_12_MSTR_RASDP_ERR	BIT(0)
+
+#define APPL_INTR_EN_L1_12			0x6c
+#define APPL_INTR_EN_L1_12_SLV_RASDP_ERR	BIT(1)
+#define APPL_INTR_EN_L1_12_MSTR_RASDP_ERR	BIT(0)
+
+#define APPL_INTR_STATUS_L1_12			0x70
+#define APPL_INTR_STATUS_L1_12_SLV_RASDP_ERR	BIT(1)
+#define APPL_INTR_STATUS_L1_12_MSTR_RASDP_ERR	BIT(0)
+
 #define APPL_INTR_STATUS_L1_13			0x74
+
 #define APPL_INTR_STATUS_L1_14			0x78
+#define APPL_INTR_STATUS_L1_14_MASK		GENMASK(29, 0)
+#define APPL_INTR_STATUS_L1_14_RETRYRAM		BIT(23)
+
 #define APPL_INTR_STATUS_L1_15			0x7C
 #define APPL_INTR_STATUS_L1_17			0x88
 
+#define APPL_FAULT_EN_L1_18				0x8c
+#define APPL_FAULT_EN_L1_18_CDM_REG_CHK_CMP_ERR		BIT(1)
+#define APPL_FAULT_EN_L1_18_CDM_REG_CHK_LOGIC_ERR	BIT(0)
+
 #define APPL_INTR_EN_L1_18				0x90
 #define APPL_INTR_EN_L1_18_CDM_REG_CHK_CMPLT		BIT(2)
 #define APPL_INTR_EN_L1_18_CDM_REG_CHK_CMP_ERR		BIT(1)
@@ -171,6 +226,19 @@
 #define APPL_CAR_RESET_OVRD				0x12C
 #define APPL_CAR_RESET_OVRD_CYA_OVERRIDE_CORE_RST_N	BIT(0)
 
+
+#define APPL_FAULT_EN_L1_20			0x188
+#define APPL_FAULT_EN_L1_20_IF_TIMEOUT		BIT(1)
+#define APPL_FAULT_EN_L1_20_SAFETY_UNCORR	BIT(0)
+
+#define APPL_INTR_EN_L1_20			0x18c
+#define APPL_INTR_EN_L1_20_IF_TIMEOUT		BIT(1)
+#define APPL_INTR_EN_L1_20_SAFETY_UNCORR	BIT(0)
+
+#define APPL_INTR_STATUS_L1_20			0x190
+#define APPL_INTR_STATUS_L1_20_IF_TIMEOUT	BIT(1)
+#define APPL_INTR_STATUS_L1_20_SAFETY_UNCORR	BIT(0)
+
 #define IO_BASE_IO_DECODE				BIT(0)
 #define IO_BASE_IO_DECODE_BIT8				BIT(8)
 
@@ -190,6 +258,12 @@
 #define AMBA_ERROR_RESPONSE_RRS_OKAY_FFFFFFFF	1
 #define AMBA_ERROR_RESPONSE_RRS_OKAY_FFFF0001	2
 
+#define PL_IF_TIMER_CONTROL_OFF			0x930
+#define PL_IF_TIMER_CONTROL_OFF_IF_TIMER_EN	BIT(0)
+#define PL_IF_TIMER_CONTROL_OFF_IF_TIMER_AER_EN	BIT(1)
+
+#define PL_INTERFACE_TIMER_STATUS_OFF		0x938
+
 #define MSIX_ADDR_MATCH_LOW_OFF			0x940
 #define MSIX_ADDR_MATCH_LOW_OFF_EN		BIT(0)
 #define MSIX_ADDR_MATCH_LOW_OFF_MASK		GENMASK(31, 2)
@@ -197,6 +271,22 @@
 #define MSIX_ADDR_MATCH_HIGH_OFF		0x944
 #define MSIX_ADDR_MATCH_HIGH_OFF_MASK		GENMASK(31, 0)
 
+#define PL_SAFETY_MASK_OFF			0x960
+#define PL_SAFETY_MASK_OFF_RASDP		BIT(0)
+#define PL_SAFETY_MASK_OFF_CDM			BIT(1)
+#define PL_SAFETY_MASK_OFF_IF_TIMEOUT		BIT(2)
+#define PL_SAFETY_MASK_OFF_UNCOR		BIT(3)
+#define PL_SAFETY_MASK_OFF_COR			BIT(4)
+#define PL_SAFETY_MASK_OFF_RASDP_COR		BIT(5)
+
+#define PL_SAFETY_STATUS_OFF			0x964
+#define PL_SAFETY_STATUS_OFF_RASDP		BIT(0)
+#define PL_SAFETY_STATUS_OFF_CDM		BIT(1)
+#define PL_SAFETY_STATUS_OFF_IF_TIMEOUT		BIT(2)
+#define PL_SAFETY_STATUS_OFF_UNCOR		BIT(3)
+#define PL_SAFETY_STATUS_OFF_COR		BIT(4)
+#define PL_SAFETY_STATUS_OFF_RASDP_COR		BIT(5)
+
 #define PORT_LOGIC_MSIX_DOORBELL			0x948
 
 #define CAP_SPCIE_CAP_OFF			0x154
@@ -229,6 +319,31 @@ static const unsigned int pcie_gen_freq[] = {
 	GEN4_CORE_CLK_FREQ
 };
 
+struct pcie_epl_error_code {
+	/* Indicates source of error */
+	uint16_t reporter_id;
+	/* Error code indicates error reported by corresponding reporter_id */
+	uint32_t error_code;
+};
+
+/*
+ * Tegra234 PCIe HSI error codes and reporter ids, refer to link
+ * https://nvidia.jamacloud.com/perspective.req#/items/22181007?projectId=22719
+ */
+static const struct pcie_epl_error_code epl_error_code[] = {
+	{ 0x8023, 0x211e },
+	{ 0x8024, 0x211f },
+	{ 0x8025, 0x2120 },
+	{ 0x8026, 0x2121 },
+	{ 0x8027, 0x2122 },
+	{ 0x8028, 0x2123 },
+	{ 0x8029, 0x2124 },
+	{ 0x802a, 0x2125 },
+	{ 0x802b, 0x2126 },
+	{ 0x802c, 0x2127 },
+	{ 0x802d, 0x212a },
+};
+
 struct tegra_pcie_dw_of_data {
 	u32 version;
 	enum dw_pcie_device_mode mode;
@@ -263,8 +378,16 @@ struct tegra_pcie_dw {
 	bool enable_cdm_check;
 	bool enable_srns;
 	bool link_state;
+	bool link_status_change;
+	bool link_speed_change;
 	bool update_fc_fixup;
 	bool enable_ext_refclk;
+	bool is_safety_platform;
+
+	atomic_t report_epl_error;
+	atomic_t bme_state_change;
+	atomic_t ep_link_up;
+
 	u8 init_link_width;
 	u32 msi_ctrl_int;
 	u32 num_lanes;
@@ -371,6 +494,181 @@ static void apply_bad_link_workaround(struct dw_pcie_rp *pp)
 	}
 }
 
+/* Read TSC counter for timestamp. */
+static inline u64 rdtsc(void)
+{
+	u64 val;
+
+	asm volatile("mrs %0, cntvct_el0" : "=r" (val));
+
+	return val;
+}
+
+static int tegra_pcie_safety_irq_handler(struct tegra_pcie_dw *pcie, u32 status_l0)
+{
+	struct dw_pcie *pci = &pcie->pci;
+	u32 val, status_l1, en_l0;
+	int irq_ret = IRQ_HANDLED;
+
+	atomic_set(&pcie->report_epl_error, 0);
+	en_l0 = appl_readl(pcie, APPL_INTR_EN_L0_0);
+
+	/* Consistency Monitor for Configuration Registers(CDM) */
+	if ((status_l0 & APPL_INTR_STATUS_L0_CDM_REG_CHK_INT) &&
+	    (en_l0 & pcie->of_data->cdm_chk_int_en_bit)) {
+		status_l1 = appl_readl(pcie, APPL_INTR_STATUS_L1_18);
+		val = dw_pcie_readl_dbi(pci, PCIE_PL_CHK_REG_CONTROL_STATUS);
+		if (status_l1 & APPL_INTR_STATUS_L1_18_CDM_REG_CHK_CMPLT) {
+			dev_info(pcie->dev, "CDM check complete\n");
+			val |= PCIE_PL_CHK_REG_CHK_REG_COMPLETE;
+		}
+		if (status_l1 & APPL_INTR_STATUS_L1_18_CDM_REG_CHK_CMP_ERR) {
+			dev_err(pcie->dev, "CDM comparison mismatch\n");
+			val |= PCIE_PL_CHK_REG_CHK_REG_COMPARISON_ERROR;
+		}
+		if (status_l1 & APPL_INTR_STATUS_L1_18_CDM_REG_CHK_LOGIC_ERR) {
+			dev_err(pcie->dev, "CDM Logic error\n");
+			val |= PCIE_PL_CHK_REG_CHK_REG_LOGIC_ERROR;
+		}
+		dw_pcie_writel_dbi(pci, PCIE_PL_CHK_REG_CONTROL_STATUS, val);
+		val = dw_pcie_readl_dbi(pci, PCIE_PL_CHK_REG_ERR_ADDR);
+		dev_err(pcie->dev, "CDM Error Address Offset = 0x%08X\n", val);
+
+		if (status_l1 & (APPL_INTR_STATUS_L1_18_CDM_REG_CHK_CMP_ERR |
+		    APPL_INTR_STATUS_L1_18_CDM_REG_CHK_LOGIC_ERR)) {
+			/*
+			 * Config space may not recover after CDM errors, disable all CDM
+			 * interrupts to avoid interrupt storm.
+			 */
+			appl_writel(pcie, 0x0, APPL_INTR_EN_L1_18);
+			appl_writel(pcie, 0x0, APPL_FAULT_EN_L1_18);
+
+			val = appl_readl(pcie, APPL_INTR_EN_L0_0);
+			val &= ~pcie->of_data->cdm_chk_int_en_bit;
+			appl_writel(pcie, val, APPL_INTR_EN_L0_0);
+
+			val = appl_readl(pcie, APPL_FAULT_EN_L0);
+			val &= ~APPL_FAULT_EN_L0_CDM_REG_CHK_FAULT_EN;
+			appl_writel(pcie, val, APPL_FAULT_EN_L0);
+
+			atomic_set(&pcie->report_epl_error, 1);
+			irq_ret = IRQ_WAKE_THREAD;
+		}
+	}
+
+	/* TLP errors like ECRC, CPL TO, etc. */
+	if ((status_l0 & APPL_INTR_STATUS_L0_TLP_ERR_INT) &&
+	    (en_l0 & APPL_INTR_EN_L0_0_TLP_ERR_INT_EN)) {
+		status_l1 = appl_readl(pcie, APPL_INTR_STATUS_L1_11);
+		appl_writel(pcie, status_l1, APPL_INTR_STATUS_L1_11);
+
+		/* Report uncorrectable errors like ECRC */
+		if (status_l1 & (APPL_INTR_STATUS_L1_11_NF_ERR_STATE |
+		    APPL_INTR_STATUS_L1_11_F_ERR_STATE)) {
+			/* Disable TLP_ERR_INT */
+			appl_writel(pcie, 0x0, APPL_INTR_EN_L1_11);
+			appl_writel(pcie, 0x0, APPL_FAULT_EN_L1_11);
+
+			val = appl_readl(pcie, APPL_INTR_EN_L0_0);
+			val &= ~APPL_INTR_EN_L0_0_TLP_ERR_INT_EN;
+			appl_writel(pcie, val, APPL_INTR_EN_L0_0);
+
+			val = appl_readl(pcie, APPL_FAULT_EN_L0);
+			val &= ~APPL_FAULT_EN_L0_TLP_ERR_FAULT_EN;
+			appl_writel(pcie, val, APPL_FAULT_EN_L0);
+
+			atomic_set(&pcie->report_epl_error, 1);
+			irq_ret = IRQ_WAKE_THREAD;
+		}
+	}
+
+	/* Uncorrectable Memory ECC errors */
+	if ((status_l0 & APPL_INTR_STATUS_L0_RASDP_INT) &&
+	    (en_l0 & APPL_INTR_EN_L0_0_RASDP_INT_EN)) {
+		status_l1 = appl_readl(pcie, APPL_INTR_STATUS_L1_12);
+
+		if (status_l1 & (APPL_INTR_STATUS_L1_12_SLV_RASDP_ERR |
+		    APPL_INTR_STATUS_L1_12_MSTR_RASDP_ERR)) {
+			/* Link is not reliable after RASDP, so disable interrupts. */
+			appl_writel(pcie, 0x0, APPL_FAULT_EN_L1_12);
+			appl_writel(pcie, 0x0, APPL_INTR_EN_L1_12);
+
+			val = appl_readl(pcie, APPL_INTR_EN_L0_0);
+			val &= ~APPL_INTR_EN_L0_0_RASDP_INT_EN;
+			appl_writel(pcie, val, APPL_INTR_EN_L0_0);
+
+			val = appl_readl(pcie, APPL_FAULT_EN_L0);
+			val &= ~APPL_FAULT_EN_L0_RASDP_FAULT_EN;
+			appl_writel(pcie, val, APPL_FAULT_EN_L0);
+
+			atomic_set(&pcie->report_epl_error, 1);
+			irq_ret = IRQ_WAKE_THREAD;
+		}
+	}
+
+	/* Parity errors */
+	if ((status_l0 & APPL_INTR_STATUS_L0_PARITY_ERR_INT) &&
+	    (en_l0 & APPL_INTR_EN_L0_0_PARITY_ERR_INT_EN)) {
+		status_l1 = appl_readl(pcie, APPL_INTR_STATUS_L1_14);
+		appl_writel(pcie, status_l1, APPL_INTR_STATUS_L1_14);
+
+		if (status_l1 & APPL_INTR_STATUS_L1_14_MASK) {
+			/* Don't report EPL error if only RETRYRAM is set */
+			if (status_l1 & ~APPL_INTR_STATUS_L1_14_RETRYRAM) {
+				/* Disable PARITY_ERR */
+				val = appl_readl(pcie, APPL_INTR_EN_L0_0);
+				val &= ~APPL_INTR_EN_L0_0_PARITY_ERR_INT_EN;
+				appl_writel(pcie, val, APPL_INTR_EN_L0_0);
+
+				val = appl_readl(pcie, APPL_FAULT_EN_L0);
+				val &= ~APPL_FAULT_EN_L0_PARITY_ERR_FAULT_EN;
+				appl_writel(pcie, val, APPL_FAULT_EN_L0);
+
+				atomic_set(&pcie->report_epl_error, 1);
+				irq_ret = IRQ_WAKE_THREAD;
+			}
+		}
+	}
+
+	/* Interface transaction timeout errors */
+	if ((status_l0 & APPL_INTR_STATUS_L0_SAFETY_UNCORR_INT) &&
+	    (en_l0 & APPL_INTR_EN_L0_0_SAFETY_UNCORR_INT_EN)) {
+		status_l1 = appl_readl(pcie, APPL_INTR_STATUS_L1_20);
+
+		/* W1C interface transaction timeout errors in PL_INTERFACE_TIMER_STATUS_OFF */
+		val = dw_pcie_readl_dbi(pci, PL_INTERFACE_TIMER_STATUS_OFF);
+		dw_pcie_writel_dbi(pci, PL_INTERFACE_TIMER_STATUS_OFF, val);
+
+		/* W1C interface transaction timeout error in PL_SAFETY_STATUS_OFF_IF_TIMEOUT */
+		val = dw_pcie_readl_dbi(pci, PL_SAFETY_STATUS_OFF);
+		dw_pcie_writel_dbi(pci, PL_SAFETY_STATUS_OFF, val);
+
+		if (status_l1 & APPL_INTR_EN_L1_20_IF_TIMEOUT) {
+			/* Disable SAFETY_UNCORR error */
+			val = appl_readl(pcie, APPL_FAULT_EN_L1_20);
+			val &= ~APPL_FAULT_EN_L1_20_IF_TIMEOUT;
+			appl_writel(pcie, val, APPL_FAULT_EN_L1_20);
+
+			val = appl_readl(pcie, APPL_INTR_EN_L1_20);
+			val &= ~APPL_INTR_EN_L1_20_IF_TIMEOUT;
+			appl_writel(pcie, val, APPL_INTR_EN_L1_20);
+
+			val = appl_readl(pcie, APPL_INTR_EN_L0_0);
+			val &= ~APPL_INTR_EN_L0_0_SAFETY_UNCORR_INT_EN;
+			appl_writel(pcie, val, APPL_INTR_EN_L0_0);
+
+			val = appl_readl(pcie, APPL_FAULT_EN_L0);
+			val &= ~APPL_FAULT_EN_L0_SAFETY_UNCORR_FAULT_EN;
+			appl_writel(pcie, val, APPL_FAULT_EN_L0);
+
+			atomic_set(&pcie->report_epl_error, 1);
+			irq_ret = IRQ_WAKE_THREAD;
+		}
+	}
+
+	return irq_ret;
+}
+
 static irqreturn_t tegra_pcie_rp_irq_handler(int irq, void *arg)
 {
 	struct tegra_pcie_dw *pcie = arg;
@@ -378,6 +676,7 @@ static irqreturn_t tegra_pcie_rp_irq_handler(int irq, void *arg)
 	struct dw_pcie_rp *pp = &pci->pp;
 	u32 val, status_l0, status_l1;
 	u16 val_w;
+	int irq_ret = IRQ_HANDLED;
 
 	status_l0 = appl_readl(pcie, APPL_INTR_STATUS_L0);
 	if (status_l0 & APPL_INTR_STATUS_L0_LINK_STATE_INT) {
@@ -398,6 +697,14 @@ static irqreturn_t tegra_pcie_rp_irq_handler(int irq, void *arg)
 			val |= PORT_LOGIC_SPEED_CHANGE;
 			dw_pcie_writel_dbi(pci, PCIE_LINK_WIDTH_SPEED_CONTROL, val);
 		}
+		if (status_l1 & APPL_INTR_STATUS_L1_0_0_RDLH_LINK_UP_CHGED) {
+			val = appl_readl(pcie, APPL_LINK_STATUS);
+			if (val & APPL_LINK_STATUS_RDLH_LINK_UP) {
+				dev_info(pcie->dev, "Link is up\n");
+				pcie->link_status_change = true;
+				irq_ret = IRQ_WAKE_THREAD;
+			}
+		}
 	}
 
 	if (status_l0 & APPL_INTR_STATUS_L0_INT_INT) {
@@ -418,37 +725,102 @@ static irqreturn_t tegra_pcie_rp_irq_handler(int irq, void *arg)
 			appl_writel(pcie,
 				    APPL_INTR_STATUS_L1_8_0_BW_MGT_INT_STS,
 				    APPL_INTR_STATUS_L1_8_0);
-
-			val_w = dw_pcie_readw_dbi(pci, pcie->pcie_cap_base +
-						  PCI_EXP_LNKSTA);
+			pcie->link_speed_change = true;
+			irq_ret = IRQ_WAKE_THREAD;
 			dev_dbg(pci->dev, "Link Speed : Gen-%u\n", val_w &
 				PCI_EXP_LNKSTA_CLS);
 		}
 	}
 
-	if (status_l0 & APPL_INTR_STATUS_L0_CDM_REG_CHK_INT) {
-		status_l1 = appl_readl(pcie, APPL_INTR_STATUS_L1_18);
-		val = dw_pcie_readl_dbi(pci, PCIE_PL_CHK_REG_CONTROL_STATUS);
-		if (status_l1 & APPL_INTR_STATUS_L1_18_CDM_REG_CHK_CMPLT) {
-			dev_info(pci->dev, "CDM check complete\n");
-			val |= PCIE_PL_CHK_REG_CHK_REG_COMPLETE;
-		}
-		if (status_l1 & APPL_INTR_STATUS_L1_18_CDM_REG_CHK_CMP_ERR) {
-			dev_err(pci->dev, "CDM comparison mismatch\n");
-			val |= PCIE_PL_CHK_REG_CHK_REG_COMPARISON_ERROR;
-		}
-		if (status_l1 & APPL_INTR_STATUS_L1_18_CDM_REG_CHK_LOGIC_ERR) {
-			dev_err(pci->dev, "CDM Logic error\n");
-			val |= PCIE_PL_CHK_REG_CHK_REG_LOGIC_ERROR;
-		}
-		dw_pcie_writel_dbi(pci, PCIE_PL_CHK_REG_CONTROL_STATUS, val);
-		val = dw_pcie_readl_dbi(pci, PCIE_PL_CHK_REG_ERR_ADDR);
-		dev_err(pci->dev, "CDM Error Address Offset = 0x%08X\n", val);
+	/* don't overwrite irq_ret if return value is not IRQ_WAKE_THREAD */
+	if (tegra_pcie_safety_irq_handler(pcie, status_l0) == IRQ_WAKE_THREAD)
+		irq_ret = IRQ_WAKE_THREAD;
+
+	return irq_ret;
+}
+
+static irqreturn_t tegra_pcie_rp_irq_thread(int irq, void *arg)
+{
+	struct tegra_pcie_dw *pcie = arg;
+	struct dw_pcie_rp *pp;
+	struct pci_bus *bus;
+	u16 speed;
+
+#if IS_ENABLED(CONFIG_TEGRA_EPL)
+	if (atomic_dec_and_test(&pcie->report_epl_error)) {
+		struct epl_error_report_frame error_report;
+		struct dw_pcie *pci = &pcie->pci;
+		int ret;
+
+		error_report.error_code = epl_error_code[pcie->cid].error_code;
+		error_report.timestamp = lower_32_bits(rdtsc());
+		error_report.reporter_id = epl_error_code[pcie->cid].reporter_id;
+
+		ret = epl_report_error(error_report);
+		if (ret < 0)
+			dev_err(pci->dev, "failed to report EPL error: %d\n", ret);
+	}
+#endif
+
+	pp = &pcie->pci.pp;
+	bus = pp->bridge->bus;
+
+	if (bus && pcie->link_status_change) {
+		pci_lock_rescan_remove();
+		pci_rescan_bus(bus);
+		pci_unlock_rescan_remove();
+	}
+
+	if (pcie->link_status_change || pcie->link_speed_change) {
+		pcie->link_status_change = false;
+		pcie->link_speed_change = false;
+		speed = dw_pcie_readw_dbi(pci,
+					  pcie->pcie_cap_base + PCI_EXP_LNKSTA);
+		speed &= PCI_EXP_LNKSTA_CLS;
+		if ((speed > 0) && (speed <= 4) && !pcie->is_safety_platform)
+			clk_set_rate(pcie->core_clk, pcie_gen_freq[speed - 1]);
 	}
 
 	return IRQ_HANDLED;
 }
 
+#if (IS_ENABLED(CONFIG_TEGRA_HSIERRRPTINJ))
+/* Error report injection test support is included */
+static int tegra_pcie_inject_err_fsi(unsigned int inst_id,
+				     struct epl_error_report_frame err_rpt_frame,
+				     void *data)
+{
+	struct tegra_pcie_dw *pcie = (struct tegra_pcie_dw *)data;
+	int err = -EINVAL;
+
+	if (!pcie)
+		return err;
+
+	/* Sanity check inst_id */
+	if (inst_id != pcie->cid) {
+		dev_err(pcie->dev, "Invalid Input -> Instance ID = 0x%04x\n", inst_id);
+		return err;
+	}
+
+	/* Sanity check reporter_id */
+	if (err_rpt_frame.reporter_id != epl_error_code[pcie->cid].reporter_id) {
+		dev_err(pcie->dev, "Invalid Input -> Reporter ID = 0x%04x\n",
+			err_rpt_frame.reporter_id);
+		return err;
+	}
+
+	/* Report error to FSI */
+	err = epl_report_error(err_rpt_frame);
+	if (err != 0) {
+		dev_err(pcie->dev, "Error injection failed for err_id: %u",
+			err_rpt_frame.error_code);
+		return err;
+	}
+
+	return err;
+}
+#endif
+
 static void pex_ep_event_hot_rst_done(struct tegra_pcie_dw *pcie)
 {
 	u32 val;
@@ -478,43 +850,57 @@ static void pex_ep_event_hot_rst_done(struct tegra_pcie_dw *pcie)
 static irqreturn_t tegra_pcie_ep_irq_thread(int irq, void *arg)
 {
 	struct tegra_pcie_dw *pcie = arg;
-	struct dw_pcie_ep *ep = &pcie->pci.ep;
 	struct dw_pcie *pci = &pcie->pci;
 	u32 val;
 
-	if (test_and_clear_bit(0, &pcie->link_status))
-		dw_pcie_ep_linkup(ep);
+#if IS_ENABLED(CONFIG_TEGRA_EPL)
+	if (atomic_dec_and_test(&pcie->report_epl_error)) {
+		struct epl_error_report_frame error_report;
+		int ret;
 
-	tegra_pcie_icc_set(pcie);
+		error_report.error_code = epl_error_code[pcie->cid].error_code;
+		error_report.timestamp = lower_32_bits(rdtsc());
+		error_report.reporter_id = epl_error_code[pcie->cid].reporter_id;
 
-	if (pcie->of_data->has_ltr_req_fix)
-		return IRQ_HANDLED;
+		ret = epl_report_error(error_report);
+		if (ret < 0)
+			dev_err(pci->dev, "failed to report EPL error: %d\n", ret);
+	}
+#endif
 
-	/* If EP doesn't advertise L1SS, just return */
-	if (!pci->l1ss_support)
-		return IRQ_HANDLED;
+	if (atomic_dec_and_test(&pcie->ep_link_up))
+		tegra_pcie_icc_set(pcie);
 
-	/* Check if BME is set to '1' */
-	val = dw_pcie_readl_dbi(pci, PCI_COMMAND);
-	if (val & PCI_COMMAND_MASTER) {
-		ktime_t timeout;
+	if (atomic_dec_and_test(&pcie->bme_state_change)) {
+		if (pcie->of_data->has_ltr_req_fix)
+			return IRQ_HANDLED;
 
-		/* Send LTR upstream */
-		val = appl_readl(pcie, APPL_LTR_MSG_2);
-		val |= APPL_LTR_MSG_2_LTR_MSG_REQ_STATE;
-		appl_writel(pcie, val, APPL_LTR_MSG_2);
+		/* If EP doesn't advertise L1SS, just return */
+		if (!pci->l1ss_support)
+			return IRQ_HANDLED;
+
+		/* Check if BME is set to '1' */
+		val = dw_pcie_readl_dbi(pci, PCI_COMMAND);
+		if (val & PCI_COMMAND_MASTER) {
+			ktime_t timeout;
 
-		timeout = ktime_add_us(ktime_get(), LTR_MSG_TIMEOUT);
-		for (;;) {
+			/* Send LTR upstream */
 			val = appl_readl(pcie, APPL_LTR_MSG_2);
-			if (!(val & APPL_LTR_MSG_2_LTR_MSG_REQ_STATE))
-				break;
-			if (ktime_after(ktime_get(), timeout))
-				break;
-			usleep_range(1000, 1100);
+			val |= APPL_LTR_MSG_2_LTR_MSG_REQ_STATE;
+			appl_writel(pcie, val, APPL_LTR_MSG_2);
+
+			timeout = ktime_add_us(ktime_get(), LTR_MSG_TIMEOUT);
+			for (;;) {
+				val = appl_readl(pcie, APPL_LTR_MSG_2);
+				if (!(val & APPL_LTR_MSG_2_LTR_MSG_REQ_STATE))
+					break;
+				if (ktime_after(ktime_get(), timeout))
+					break;
+				usleep_range(1000, 1100);
+			}
+			if (val & APPL_LTR_MSG_2_LTR_MSG_REQ_STATE)
+				dev_err(pcie->dev, "Failed to send LTR message\n");
 		}
-		if (val & APPL_LTR_MSG_2_LTR_MSG_REQ_STATE)
-			dev_err(pcie->dev, "Failed to send LTR message\n");
 	}
 
 	return IRQ_HANDLED;
@@ -524,8 +910,13 @@ static irqreturn_t tegra_pcie_ep_hard_irq(int irq, void *arg)
 {
 	struct tegra_pcie_dw *pcie = arg;
 	int spurious = 1;
+	struct dw_pcie_ep *ep = &pcie->pci.ep;
+	int irq_ret = IRQ_HANDLED;
 	u32 status_l0, status_l1, link_status;
 
+	atomic_set(&pcie->ep_link_up, 0);
+	atomic_set(&pcie->bme_state_change, 0);
+
 	status_l0 = appl_readl(pcie, APPL_INTR_STATUS_L0);
 	if (status_l0 & APPL_INTR_STATUS_L0_LINK_STATE_INT) {
 		status_l1 = appl_readl(pcie, APPL_INTR_STATUS_L1_0_0);
@@ -538,8 +929,9 @@ static irqreturn_t tegra_pcie_ep_hard_irq(int irq, void *arg)
 			link_status = appl_readl(pcie, APPL_LINK_STATUS);
 			if (link_status & APPL_LINK_STATUS_RDLH_LINK_UP) {
 				dev_dbg(pcie->dev, "Link is up with Host\n");
-				set_bit(0, &pcie->link_status);
-				return IRQ_WAKE_THREAD;
+				dw_pcie_ep_linkup(ep);
+				atomic_set(&pcie->ep_link_up, 1);
+				irq_ret = IRQ_WAKE_THREAD;
 			}
 		}
 
@@ -550,9 +942,10 @@ static irqreturn_t tegra_pcie_ep_hard_irq(int irq, void *arg)
 		status_l1 = appl_readl(pcie, APPL_INTR_STATUS_L1_15);
 		appl_writel(pcie, status_l1, APPL_INTR_STATUS_L1_15);
 
-		if (status_l1 & APPL_INTR_STATUS_L1_15_CFG_BME_CHGED)
-			return IRQ_WAKE_THREAD;
-
+		if (status_l1 & APPL_INTR_STATUS_L1_15_CFG_BME_CHGED) {
+			atomic_set(&pcie->bme_state_change, 1);
+			irq_ret = IRQ_WAKE_THREAD;
+		}
 		spurious = 0;
 	}
 
@@ -567,13 +960,17 @@ static irqreturn_t tegra_pcie_ep_hard_irq(int irq, void *arg)
 			spurious = 0;
 	}
 
+	/* don't overwrite irq_ret if return value is not IRQ_WAKE_THREAD*/
+	if (tegra_pcie_safety_irq_handler(pcie, status_l0) == IRQ_WAKE_THREAD)
+		irq_ret = IRQ_WAKE_THREAD;
+
 	if (spurious) {
 		dev_warn(pcie->dev, "Random interrupt (STATUS = 0x%08X)\n",
 			 status_l0);
 		appl_writel(pcie, status_l0, APPL_INTR_STATUS_L0);
 	}
 
-	return IRQ_HANDLED;
+	return irq_ret;
 }
 
 static int tegra_pcie_dw_rd_own_conf(struct pci_bus *bus, u32 devfn, int where,
@@ -750,6 +1147,67 @@ static inline void init_host_aspm(struct tegra_pcie_dw *pcie) { return; }
 static inline void init_debugfs(struct tegra_pcie_dw *pcie) { return; }
 #endif
 
+static void tegra_pcie_enable_fault_interrupts(struct tegra_pcie_dw *pcie)
+{
+	struct dw_pcie *pci = &pcie->pci;
+	u32 val;
+
+	val = appl_readl(pcie, APPL_FAULT_EN_L0);
+	val |= APPL_FAULT_EN_L0_TLP_ERR_FAULT_EN;
+	val |= APPL_FAULT_EN_L0_RASDP_FAULT_EN;
+	val |= APPL_FAULT_EN_L0_PARITY_ERR_FAULT_EN;
+	val |= APPL_FAULT_EN_L0_SAFETY_UNCORR_FAULT_EN;
+	appl_writel(pcie, val, APPL_FAULT_EN_L0);
+
+	val = appl_readl(pcie, APPL_INTR_EN_L0_0);
+	val |= APPL_INTR_EN_L0_0_TLP_ERR_INT_EN;
+	val |= APPL_INTR_EN_L0_0_RASDP_INT_EN;
+	val |= APPL_INTR_EN_L0_0_PARITY_ERR_INT_EN;
+	val |= APPL_INTR_EN_L0_0_SAFETY_UNCORR_INT_EN;
+	appl_writel(pcie, val, APPL_INTR_EN_L0_0);
+
+	/* Enable correctable errors reporting */
+	val = appl_readl(pcie, APPL_FAULT_EN_L1_11);
+	val |= (APPL_FAULT_EN_L1_11_NF_ERR_FAULT_EN | APPL_FAULT_EN_L1_11_F_ERR_FAULT_EN);
+	appl_writel(pcie, val, APPL_FAULT_EN_L1_11);
+
+	val = appl_readl(pcie, APPL_INTR_EN_L1_11);
+	val |= (APPL_INTR_EN_L1_11_NF_ERR_INT_EN | APPL_INTR_EN_L1_11_F_ERR_INT_EN);
+	appl_writel(pcie, val, APPL_INTR_EN_L1_11);
+
+	/* Enable uncorrectable memory ECC */
+	val = appl_readl(pcie, APPL_FAULT_EN_L1_12);
+	val |= APPL_FAULT_EN_L1_12_SLV_RASDP_ERR;
+	val |= APPL_FAULT_EN_L1_12_MSTR_RASDP_ERR;
+	appl_writel(pcie, val, APPL_FAULT_EN_L1_12);
+
+	val = appl_readl(pcie, APPL_INTR_EN_L1_12);
+	val |= APPL_INTR_EN_L1_12_SLV_RASDP_ERR;
+	val |= APPL_INTR_EN_L1_12_MSTR_RASDP_ERR;
+	appl_writel(pcie, val, APPL_INTR_EN_L1_12);
+
+	/* Enable interface transaction timeout */
+	val = appl_readl(pcie, APPL_FAULT_EN_L1_20);
+	val |= APPL_FAULT_EN_L1_20_IF_TIMEOUT;
+	appl_writel(pcie, val, APPL_FAULT_EN_L1_20);
+
+	val = appl_readl(pcie, APPL_INTR_EN_L1_20);
+	val |= APPL_INTR_EN_L1_20_IF_TIMEOUT;
+	appl_writel(pcie, val, APPL_INTR_EN_L1_20);
+
+	val = dw_pcie_readl_dbi(pci, PL_IF_TIMER_CONTROL_OFF);
+	val |= PL_IF_TIMER_CONTROL_OFF_IF_TIMER_EN |
+		PL_IF_TIMER_CONTROL_OFF_IF_TIMER_AER_EN;
+	dw_pcie_writel_dbi(pci, PL_SAFETY_MASK_OFF, val);
+
+	/* Mask all uncorectable error except transaction timeout */
+	val = dw_pcie_readl_dbi(pci, PL_SAFETY_MASK_OFF);
+	val |= (PL_SAFETY_MASK_OFF_RASDP | PL_SAFETY_MASK_OFF_CDM |
+			PL_SAFETY_MASK_OFF_UNCOR | PL_SAFETY_MASK_OFF_COR |
+			PL_SAFETY_MASK_OFF_RASDP_COR);
+	dw_pcie_writel_dbi(pci, PL_SAFETY_MASK_OFF, val);
+}
+
 static void tegra_pcie_enable_system_interrupts(struct dw_pcie_rp *pp)
 {
 	struct dw_pcie *pci = to_dw_pcie_from_pp(pp);
@@ -772,12 +1230,24 @@ static void tegra_pcie_enable_system_interrupts(struct dw_pcie_rp *pp)
 		val |= pcie->of_data->cdm_chk_int_en_bit;
 		appl_writel(pcie, val, APPL_INTR_EN_L0_0);
 
+		val = appl_readl(pcie, APPL_FAULT_EN_L0);
+		val |= APPL_FAULT_EN_L0_CDM_REG_CHK_FAULT_EN;
+		appl_writel(pcie, val, APPL_FAULT_EN_L0);
+
 		val = appl_readl(pcie, APPL_INTR_EN_L1_18);
 		val |= APPL_INTR_EN_L1_18_CDM_REG_CHK_CMP_ERR;
 		val |= APPL_INTR_EN_L1_18_CDM_REG_CHK_LOGIC_ERR;
 		appl_writel(pcie, val, APPL_INTR_EN_L1_18);
+
+		val = appl_readl(pcie, APPL_FAULT_EN_L1_18);
+		val |= APPL_FAULT_EN_L1_18_CDM_REG_CHK_CMP_ERR;
+		val |= APPL_FAULT_EN_L1_18_CDM_REG_CHK_LOGIC_ERR;
+		appl_writel(pcie, val, APPL_FAULT_EN_L1_18);
 	}
 
+	if (pcie->is_safety_platform)
+		tegra_pcie_enable_fault_interrupts(pcie);
+
 	val_w = dw_pcie_readw_dbi(&pcie->pci, pcie->pcie_cap_base +
 				  PCI_EXP_LNKSTA);
 	pcie->init_link_width = FIELD_GET(PCI_EXP_LNKSTA_NLW, val_w);
@@ -1223,6 +1693,9 @@ static int tegra_pcie_dw_parse_dt(struct tegra_pcie_dw *pcie)
 	pcie->enable_cdm_check =
 		of_property_read_bool(np, "snps,enable-cdm-check");
 
+	pcie->is_safety_platform =
+		of_property_read_bool(np, "nvidia,enable-safety");
+
 	if (pcie->of_data->version == TEGRA234_DWC_IP_VER)
 		pcie->enable_srns =
 			of_property_read_bool(np, "nvidia,enable-srns");
@@ -1923,6 +2396,26 @@ static void pex_ep_event_pex_rst_deassert(struct tegra_pcie_dw *pcie)
 	val |= APPL_INTR_EN_L1_8_EDMA_INT_EN;
 	appl_writel(pcie, val, APPL_INTR_EN_L1_8_0);
 
+	if (pcie->enable_cdm_check) {
+		val = appl_readl(pcie, APPL_INTR_EN_L0_0);
+		val |= pcie->of_data->cdm_chk_int_en_bit;
+		appl_writel(pcie, val, APPL_INTR_EN_L0_0);
+
+		val = appl_readl(pcie, APPL_FAULT_EN_L0);
+		val |= APPL_FAULT_EN_L0_CDM_REG_CHK_FAULT_EN;
+		appl_writel(pcie, val, APPL_FAULT_EN_L0);
+
+		val = appl_readl(pcie, APPL_INTR_EN_L1_18);
+		val |= APPL_INTR_EN_L1_18_CDM_REG_CHK_CMP_ERR;
+		val |= APPL_INTR_EN_L1_18_CDM_REG_CHK_LOGIC_ERR;
+		appl_writel(pcie, val, APPL_INTR_EN_L1_18);
+
+		val = appl_readl(pcie, APPL_FAULT_EN_L1_18);
+		val |= APPL_FAULT_EN_L1_18_CDM_REG_CHK_CMP_ERR;
+		val |= APPL_FAULT_EN_L1_18_CDM_REG_CHK_LOGIC_ERR;
+		appl_writel(pcie, val, APPL_FAULT_EN_L1_18);
+	}
+
 	/* 110us for both snoop and no-snoop */
 	val = FIELD_PREP(PCI_LTR_VALUE_MASK, 110) |
 	      FIELD_PREP(PCI_LTR_SCALE_MASK, 2) |
@@ -1934,6 +2427,9 @@ static void pex_ep_event_pex_rst_deassert(struct tegra_pcie_dw *pcie)
 
 	reset_control_deassert(pcie->core_rst);
 
+	if (pcie->is_safety_platform)
+		tegra_pcie_enable_fault_interrupts(pcie);
+
 	/* Perform cleanup that requires refclk and core reset deasserted */
 	pci_epc_deinit_notify(pcie->pci.ep.epc);
 	dw_pcie_ep_cleanup(&pcie->pci.ep);
@@ -2411,10 +2907,23 @@ static int tegra_pcie_dw_probe(struct platform_device *pdev)
 		return ret;
 	}
 
+#if (IS_ENABLED(CONFIG_TEGRA_HSIERRRPTINJ))
+	/*
+	 * Continue init despite err inj utility registration failure, as the err inj support
+	 * is meant only for debug purposes.
+	 */
+	ret = hsierrrpt_reg_cb(IP_PCIE, pcie->cid, tegra_pcie_inject_err_fsi, pcie);
+	if (ret != 0)
+		dev_info(pcie->dev, "Err inj callback registration failed: %d", ret);
+	ret = 0;
+#endif
+
 	switch (pcie->of_data->mode) {
 	case DW_PCIE_RC_TYPE:
-		ret = devm_request_irq(dev, pp->irq, tegra_pcie_rp_irq_handler,
-				       IRQF_SHARED, "tegra-pcie-intr", pcie);
+		ret = devm_request_threaded_irq(dev, pp->irq, tegra_pcie_rp_irq_handler,
+						tegra_pcie_rp_irq_thread,
+						IRQF_SHARED,
+						"tegra-pcie-intr", (void *)pcie);
 		if (ret) {
 			dev_err(dev, "Failed to request IRQ %d: %d\n", pp->irq,
 				ret);
@@ -2496,6 +3005,9 @@ static int tegra_pcie_dw_probe(struct platform_device *pdev)
 	}
 
 fail:
+#if (IS_ENABLED(CONFIG_TEGRA_HSIERRRPTINJ))
+	hsierrrpt_dereg_cb(IP_PCIE, pcie->cid);
+#endif
 	tegra_bpmp_put(pcie->bpmp);
 	return ret;
 }
@@ -2504,6 +3016,12 @@ static void tegra_pcie_dw_remove(struct platform_device *pdev)
 {
 	struct tegra_pcie_dw *pcie = platform_get_drvdata(pdev);
 	struct dw_pcie_ep *ep = &pcie->pci.ep;
+#if (IS_ENABLED(CONFIG_TEGRA_HSIERRRPTINJ))
+	int err;
+	err = hsierrrpt_dereg_cb(IP_PCIE, pcie->cid);
+	if (err != 0)
+		dev_info(pcie->dev, "Err inj callback de-registration failed: %d", err);
+#endif
 
 	if (pcie->of_data->mode == DW_PCIE_RC_TYPE) {
 		disable_irq(pcie->prsnt_irq);
-- 
2.43.0

